python_modules/libraries/dagster-dg/dagster_dg/cli/plus.py

```diff
@@ -142,3 +151,169 @@ def pull_env_command(**global_options: object) -> None:
+
+    location_suffix = "" if global_ else f" for location {dg_context.project_name}"
+    scope_text = f" in {', '.join(sorted(active_scopes))} scope"
+
+    existing_secrets = gql_client.execute(
+        gql.GET_SECRETS_FOR_SCOPES_QUERY,
+        variables={
+            "locationName": None if global_ else dg_context.project_name,
+            "scopes": {
+                "fullDeploymentScope": Scope.FULL in active_scopes,
+                "allBranchDeploymentsScope": Scope.BRANCH in active_scopes,
+                "localDeploymentScope": Scope.LOCAL in active_scopes,
+            },
+            "secretName": env_name,
+        },
+    )["secretsForScopes"]["secrets"]
+
+    for existing_secret in existing_secrets:
+        if len(existing_secret["locationNames"]) > 1:
+            raise click.ClickException(
+                f"Environment variable {env_name} is configured for multiple locations {', '.join(existing_secret['locationNames'])}, and cannot be modified via the CLI."
+            )
+
+    should_confirm = False
+    any_secret_is_global = any(
+        _secret_is_global(existing_secret) for existing_secret in existing_secrets
+    )
+    if not global_ and any_secret_is_global:
+        click.echo(
+            f"Environment variable {env_name} is set globally within the current deployment. The current command will only update the value for location {dg_context.project_name}. Use --global to update this value for all locations."
+        )
+        should_confirm = True
+
+    existing_scopes = set(
+        itertools.chain.from_iterable(
+            _get_secret_scopes(existing_secret)
+            for existing_secret in existing_secrets
+            if global_ or existing_secret["locationNames"]
+        )
+    )
+    if existing_scopes:
+        click.echo(
+            f"Environment variable {env_name} is already configured for {', '.join(s.value for s in existing_scopes)} scope{location_suffix}."
+        )
+        should_confirm = True
+
+    if should_confirm:
+        if not global_:
+            click.confirm(
+                f"\nAre you sure you want to update environment variable {env_name}{scope_text}{location_suffix}?",
+                abort=True,
+            )
+        else:
+            click.confirm(
+                f"\nAre you sure you want to update environment variable {env_name}{scope_text} for all locations?",
+                abort=True,
+            )
+
+    gql_client.execute(
+        gql.CREATE_OR_UPDATE_SECRET_FOR_SCOPES_MUTATION,
+        variables={
+            "secretName": env_name,
+            "secretValue": env_value,
+            "scopes": {
+                "fullDeploymentScope": Scope.FULL in active_scopes,
+                "allBranchDeploymentsScope": Scope.BRANCH in active_scopes,
+                "localDeploymentScope": Scope.LOCAL in active_scopes,
+            },
+            "locationName": None if global_ else dg_context.project_name,
+        },
+    )
```

Comment:
so the details of what we confirm here on the client seems very tied up with the specifics of what happens in this serverside mutation. Given that we're going to offer pretty clear confirmation steps for what secrets are going to updated/deleted here, should we just make the client then carry out those mutations using the existing CRUD mutations, so that we can be confident that what you confirmed matches what actually gets carried out? Instead of putting the complicated logic on the server? wdyt?

Reply:
Hmm, I could see us wanting this GQL mutation behavior elsewhere (e.g. in the UI, with a similar warn/confirmation process). I think the risk of race condition here is pretty low and the benefit is reduced complexity in the CLI itself, but don't have strong opinion here.